### PR TITLE
Add high-level CSV Subgraph

### PR DESCRIPTION
**Pull request by gforsyth**
Very much still a work-in-progress -- no optimizations applied yet, but CSV reading works as expected (some test failures because of `read_table` stuff and also some non-deterministic task naming I need to fix).

This was a little hairier than I expected since sticking `delayed` objects inside of other subgraphs was causing problems like `Delayed objects of unspecified length have no len()`.  Or at least I didn't get it to work the first time around, so this instead has switched a few things to just construct the graph directly ( and one particularly heinous hack of unpacking a delayed graph).

Definitely not ready for review, but throwing it up here so I don't forget to come back to it.

cc @TomAugspurger 

- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -243,45 +322,16 @@ def text_blocks_to_pandas(
...
-            dtypes,
-            columns,
-            write_header=False,
-            enforce=enforce,
-            path=path_info,
-        )
+    blocks = tuple(flatten(block_lists))
+    # Create mask of first blocks from nested block_lists
+    for sublist in block_lists:
+        sublist[0] = True
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply dask/dataframe/io/csv.py:329]: I think this inplace mutation is failing `dask/dataframe/io/tests/test_csv.py::test_text_blocks_to_pandas_blocked[read_csv-files0]`.

File: dask/dataframe/io/csv.py [hunk 2]
```diff
@@ -17,17 +17,96 @@
...
+        else:
+            path_info = None
+
+        write_header = False
+        rest_kwargs = self.kwargs.copy()
+        if not self.is_first[i]:
+            write_header = True
+            rest_kwargs.pop("skiprows", None)
+
+        return (pandas_read_text,
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py:90]: Seems to be a few failures related to the changed result type of `text_block_to_pandas`. I view this as an internal function so I think it's OK to just update the tests.

**Comment by gforsyth** [comment]
> Anything I can do to help unstick this @gforsyth?

Hey @TomAugspurger -- thanks for the review! I was just getting back to work on this, I'll remove the in-place mutation to start with.

I started another version of this where `text_block_to_pandas` dispatches based on the value of `collections`, thinking to separate out the `table` behavior from the `csv` behavior.  I'm not entirely sure if that makes sense.  It leads to a little bit of code duplication.  

Is there value in maintaining the return of a list of delayed objects as an option?  Or is a `dask.dataframe` sufficient?

**Comment by TomAugspurger** [comment]
Offhand, I don't believe there's much value in the list of delayed objects.

**Comment by gforsyth** [comment]
Hey @TomAugspurger -- this is ready for another look.  All tests passing (with a few small modifications to reflect the removal of list of Delayeds).  This is just the subgraph, no optimizations applied but I think that can be done separately.

File: dask/dataframe/io/csv.py [hunk 3]
```diff
@@ -294,9 +345,31 @@ def text_blocks_to_pandas(
...
+        columns,
+        collection,
+        enforce,
+        path,
+    )
+
+    return new_dd_object(subgraph, name, head, (None,) * (len(blocks) + 1))
+
+
+def block_mask(block_lists):
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py:367]: A small docstring here would be nice.

File: dask/dataframe/io/csv.py [hunk 4]
```diff
@@ -294,9 +345,31 @@ def text_blocks_to_pandas(
...
+    for block in block_lists:
+        if not block:
+            continue
+        yield True
+        yield from (False for _ in block[1:])
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py:372]: Nice use of `yield from` :)

**Comment by gforsyth** [comment]
> Is the parameter `collection=False` still useful in `text_block_to_pandas`? My understanding is that that's what controlled the DataFrame vs. list[DataFrame] return value.

I don't think so, so I've removed it, but I can revert that commit if there's a reason to keep it.

**Comment by TomAugspurger** [comment]
@gforsyth Travis showed up, linting error: https://travis-ci.org/github/dask/dask/jobs/696500285#L1477

FYI, pre-commit is quite nice :) https://docs.dask.org/en/latest/develop.html#code-formatting

**Comment by gforsyth** [comment]
> FYI, pre-commit is quite nice

Thanks, Tom!  Had it installed, but had neglected to actually, you know, set it up on `dask/dask`.

Fixed and pushed.

**Comment by TomAugspurger** [comment]
Thanks!

**Comment by mrocklin** [comment]
Woot!

**Pull request merged**